File: app/android/app/src/main/java/com/sarvamula/reader/SvPrint.java
package com.sarvamula.reader;

import android.content.Context;
import android.print.PrintAttributes;
import android.print.PrintDocumentAdapter;
import android.print.PrintManager;
import android.webkit.WebView;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Hands the WebView to Android's print framework, whose dialog offers "Save as PDF".
 *
 * The reader's PDF buttons called window.print(), which Mobile Safari and desktop browsers
 * implement but Android WebView and iOS WKWebView do NOT — so in the apps the buttons did
 * nothing at all, silently, while working fine on the web.
 *
 * Printing goes through the platform rather than a JS PDF library on purpose: jsPDF and
 * pdfmake place glyphs with no shaping engine, so Devanagari conjuncts and mātrās break —
 * unacceptable for a critical edition. The WebView's own text stack shapes correctly, and
 * createPrintDocumentAdapter applies the page's @media print rules, so #printroot and the
 * existing print stylesheet are used unchanged.
 */
@CapacitorPlugin(name = "SvPrint")
public class SvPrint extends Plugin {

    @PluginMethod
    public void printDoc(final PluginCall call) {
        final String jobName = call.getString("name", "Sarvamula");
        // must run on the UI thread: the adapter is created from the WebView
        getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    WebView web = getBridge().getWebView();
                    if (web == null) {
                        call.reject("no webview");
                        return;
                    }
                    PrintManager pm = (PrintManager)
                            getActivity().getSystemService(Context.PRINT_SERVICE);
                    if (pm == null) {
                        call.reject("printing unavailable");
                        return;
                    }
                    PrintDocumentAdapter adapter = web.createPrintDocumentAdapter(jobName);
                    pm.print(jobName, adapter, new PrintAttributes.Builder().build());
                    call.resolve();
                } catch (Exception e) {
                    call.reject(e.getMessage(), e);
                }
            }
        });
    }
}

File: app/ios/App/App/SvPrint.swift
import Foundation
import UIKit
import WebKit
import Capacitor

/// Hands the WKWebView to iOS's print system, whose sheet offers "Save to Files" / AirPrint.
///
/// The reader's PDF buttons called window.print(), which Mobile Safari implements but
/// WKWebView inside an app does not — so the buttons did nothing at all in the app while
/// working on the web.
///
/// `viewPrintFormatter()` is used rather than `UIMarkupTextPrintFormatter`: the markup
/// formatter renders the HTML without the bundle's woff2 faces, which breaks Devanagari
/// shaping — the exact failure a JS PDF library would cause and the reason printing goes
/// through the platform at all. The view formatter prints what the WebView has already laid
/// out, so #printroot and the @media print rules apply unchanged.
@objc(SvPrint)
public class SvPrint: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SvPrint"
    public let jsName = "SvPrint"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "printDoc", returnType: CAPPluginReturnPromise)
    ]

    // NOT named `print`: that would shadow Swift's global print(_:) inside this type.
    @objc func printDoc(_ call: CAPPluginCall) {
        let jobName = call.getString("name") ?? "Sarvamūla"
        DispatchQueue.main.async {
            guard let web = self.bridge?.webView else {
                call.reject("no webview")
                return
            }
            let info = UIPrintInfo(dictionary: nil)
            info.outputType = .general
            info.jobName = jobName
            let controller = UIPrintInteractionController.shared
            controller.printInfo = info
            controller.printFormatter = web.viewPrintFormatter()
            controller.present(animated: true) { (_, completed, error) in
                if let error = error {
                    call.reject(error.localizedDescription)
                } else {
                    // `completed` is false when the sheet is dismissed without printing —
                    // not an error, and the caller uses it only to clear the print layout.
                    call.resolve(["completed": completed])
                }
            }
        }
    }
}
